pkg/agent/controller/noderoute/node_route_controller.go

```diff
@@ -206,67 +224,203 @@ func (c *Controller) syncNodeRoute(nodeName string) error {
 	// methods.
 
 	if node, err := c.nodeLister.Get(nodeName); err != nil {
-		klog.Infof("Deleting routes and flow entries to Node %s", nodeName)
-		route, flowsAreInstalled := c.installedNodes.Load(nodeName)
-		if route != nil {
-			if err = netlink.RouteDel(route.(*netlink.Route)); err != nil {
-				return fmt.Errorf("failed to delete the route to Node %s: %v", nodeName, err)
-			}
-			c.installedNodes.Store(nodeName, nil)
+		return c.deleteNodeRoute(nodeName)
+	} else {
+		return c.addNodeRoute(nodeName, node)
+	}
+}
+
+func (c *Controller) deleteNodeRoute(nodeName string) error {
+	klog.Infof("Deleting routes and flows to Node %s", nodeName)
+
+	route, flowsAreInstalled := c.installedNodes.Load(nodeName)
+	if route != nil {
+		if err := netlink.RouteDel(route.(*netlink.Route)); err != nil {
+			return fmt.Errorf("failed to delete the route to Node %s: %v", nodeName, err)
 		}
-		if flowsAreInstalled {
-			if err = c.ofClient.UninstallNodeFlows(nodeName); err != nil {
-				return fmt.Errorf("failed to uninstall flows to Node %s: %v", nodeName, err)
-			}
+		c.installedNodes.Store(nodeName, nil)
+	}
+
+	if flowsAreInstalled {
+		if err := c.ofClient.UninstallNodeFlows(nodeName); err != nil {
+			return fmt.Errorf("failed to uninstall flows to Node %s: %v", nodeName, err)
 		}
 		c.installedNodes.Delete(nodeName)
-	} else if route, flowsAreInstalled := c.installedNodes.Load(nodeName); route == nil {
-		klog.Infof("Adding routes and flows to Node %s, podCIDR: %s, addresses: %v",
-			nodeName, node.Spec.PodCIDR, node.Status.Addresses)
-		if node.Spec.PodCIDR == "" {
-			klog.V(1).Infof("PodCIDR is empty for peer node %s", nodeName)
+	}
+
+	if c.ipsecPSK != "" {
+		interfaceConfig, ok := c.interfaceStore.GetNodeTunnelInterface(nodeName)
+		if !ok {
+			// Tunnel port not created for this Node.
 			return nil
 		}
+		if err := c.ovsBridgeClient.DeletePort(interfaceConfig.PortUUID); err != nil {
+			klog.Errorf("Failed to delete OVS tunnel port %s for Node %s: %v",
+				interfaceConfig.InterfaceName, nodeName, err)
+			return fmt.Errorf("failed to delete OVS tunnel port for Node %s", nodeName)
```

Comment:
This is not unified before, but want to discuss it here, finally these errors will be logged in `processNextWorkItem`:
```
klog.Errorf("Error syncing Node %s, requeuing. Error: %v", key, err)
```
Do you think we can remove all `klog.Errorf` to avoid duplicate error logs? I'm thinking maybe we should only use `klog.Errorf` when the code is going to swallow the error and not pop up it, what do you think?

Reply:
Yes, I thought about it too. Just feel might be easier to debug if log closer to where the failure happens, and probably log with more details there.
For example, if we do not log here, should we always append the previously returned error to the new error and return to the caller?

Reply:
Yes, if not log here, previously returned error should be appended, like L246 (I think it's the most common way to return error in Go). That would guarantee no information will be missing. I raised it because there were only little difference between the error logs and the returned errors, while there should be no reason to not include details in returned errors too.

Reply:
Ok. Here I still feel easier to debug if log closer to the failure. But no strong opinion. If you believe we should always log in the upper layer, I could change.

Reply:
Just personal opinion, not strong too. Feel feel to ignore.